rotkehlchen/externalapis/alchemy.py

```diff
@@ -0,0 +1,400 @@
+
+                for price in prices:
+                    if price['currency'] != 'usd':
+                        continue
+
+                    return deserialize_price(price['value'])
+        except (DeserializationError, KeyError) as e:
+            error_msg = f'missing key {e!s}' if isinstance(e, KeyError) else str(e)
+            log.warning(
+                f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}. '  # noqa: E501
+                f'Failed to deserialize price data due to: {error_msg}',
+            )
+            return ZERO_PRICE
+
+        log.warning(
+            f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}, '
+            f'but no USD price entry was found in the response',
+        )
+        return ZERO_PRICE
+
+    def _deserialize_historical_price(
+            self,
+            result: dict[str, Any],
+            timestamp: Timestamp,
+            asset_identifier: dict[str, str],
+            from_asset: Asset,
+            to_asset: Asset,
+    ) -> Price:
+        """Deserializes historical price data from Alchemy API response.
+
+        If price data is missing, invalid, or there's an error, ZERO_PRICE is returned.
+        Otherwise, the deserialized price is returned.
+        """
+        try:
+            if self._is_identifier_match(result, asset_identifier) is False:
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received mismatched symbol in response',
+                )
+                return ZERO_PRICE
+
+            if result['currency'] != 'usd':
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received non-USD currency in response',
+                )
+                return ZERO_PRICE
+
+            if not (data := result.get('data', [])):
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received no historical price data in response',
+                )
+                return ZERO_PRICE
+
+            closest_price = min(
+                data,
+                key=lambda x: abs(timestamp - int(datetime.fromisoformat(x['timestamp']).timestamp())),  # noqa: E501
+            )
+            return deserialize_price(closest_price['value'])
+        except (DeserializationError, KeyError, ValueError) as e:
+            error_msg = f'missing key {e!s}' if isinstance(e, KeyError) else str(e)
+            log.warning(
+                f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}. '  # noqa: E501
+                f'Failed to deserialize historical price data due to: {error_msg}',
+            )
+            return ZERO_PRICE
+
+    @staticmethod
+    def _is_identifier_match(
```

Comment:
This should also be a function of the `AlchemyAssetData` named tuple / data class I said you should make.